File: examples/go/inventory/logic/initialize_stock.go
package logic

import (
	"inventory/proto/examples"

	"google.golang.org/protobuf/types/known/timestamppb"
)

func (l *DefaultInventoryLogic) HandleInitializeStock(state *InventoryState, productID string, quantity, lowStockThreshold int32) (*examples.StockInitialized, error) {
	if state.Exists() {
		return nil, NewFailedPrecondition("Inventory already initialized")
	}
	if productID == "" {
		return nil, NewInvalidArgument("Product ID is required")
	}
	if quantity < 0 {
		return nil, NewInvalidArgument("Quantity cannot be negative")
	}
	if lowStockThreshold < 0 {
		return nil, NewInvalidArgument("Low stock threshold cannot be negative")
	}

	return &examples.StockInitialized{
		ProductId:         productID,
		Quantity:          quantity,
		LowStockThreshold: lowStockThreshold,
		InitializedAt:     timestamppb.Now(),
	}, nil
}

File: examples/python/agg-inventory/handlers/initialize_stock.py
"""Handler for InitializeStock command."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import types_pb2 as types
from errors import CommandRejectedError
from proto import inventory_pb2 as inventory

from handlers.state import InventoryState


def handle_initialize_stock(
    command_book: types.CommandBook,
    command_any: Any,
    state: InventoryState,
    seq: int,
) -> types.EventBook:
    """Handle InitializeStock command."""
    if state.exists():
        raise CommandRejectedError("Inventory already initialized")

    cmd = inventory.InitializeStock()
    command_any.Unpack(cmd)

    if not cmd.product_id:
        raise CommandRejectedError("Product ID is required")
    if cmd.quantity < 0:
        raise CommandRejectedError("Quantity cannot be negative")
    if cmd.low_stock_threshold < 0:
        raise CommandRejectedError("Low stock threshold cannot be negative")

    event = inventory.StockInitialized(
        product_id=cmd.product_id,
        quantity=cmd.quantity,
        low_stock_threshold=cmd.low_stock_threshold,
        initialized_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return types.EventBook(
        cover=command_book.cover,
        pages=[
            types.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )
